src/main/java/org/kohsuke/github/extras/authorization/JWTTokenProvider.java

```diff
@@ -0,0 +1,130 @@
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+import org.kohsuke.github.authorization.AuthorizationProvider;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.GeneralSecurityException;
+import java.security.KeyFactory;
+import java.security.PrivateKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.Date;
+
+import javax.annotation.Nonnull;
+
+/**
+ * A authorization provider that gives valid JWT tokens. These tokens are then used to create a time-based token to
+ * authenticate as an application. This token provider does not provide any kind of caching, and will always request a
+ * new token to the API.
+ */
+public class JWTTokenProvider implements AuthorizationProvider {
+
+    private final PrivateKey privateKey;
+
+    @Nonnull
+    private Instant validUntil = Instant.MIN;
+
+    private String token;
+
+    /**
+     * The identifier for the application
+     */
+    private final String applicationId;
+
+    public JWTTokenProvider(String applicationId, File keyFile) throws GeneralSecurityException, IOException {
+        this(applicationId, loadPrivateKey(keyFile.toPath()));
+    }
+
+    public JWTTokenProvider(String applicationId, Path keyPath) throws GeneralSecurityException, IOException {
+        this(applicationId, loadPrivateKey(keyPath));
+    }
+
+    public JWTTokenProvider(String applicationId, PrivateKey privateKey) {
+        this.privateKey = privateKey;
+        this.applicationId = applicationId;
+    }
+
+    @Override
+    public String getEncodedAuthorization() throws IOException {
+        synchronized (this) {
+            if (Instant.now().isAfter(validUntil)) {
+                token = refreshJWT();
+            }
+            return String.format("Bearer %s", token);
+        }
+    }
+
+    /**
+     * add dependencies for a jwt suite You can generate a key to load in this method with:
+     *
+     * <pre>
+     * openssl pkcs8 -topk8 -inform PEM -outform DER -in ~/github-api-app.private-key.pem -out ~/github-api-app.private-key.der -nocrypt
+     * </pre>
+     */
+    private static PrivateKey loadPrivateKey(Path keyPath) throws GeneralSecurityException, IOException {
+        String keyString = new String(Files.readAllBytes(keyPath), StandardCharsets.UTF_8);
+        return getPrivateKeyFromString(keyString);
+    }
+
+    /**
+     * Convert a PKCS#8 formatted private key in string format into a java PrivateKey
+     *
+     * @param key
+     *            PCKS#8 string
+     * @return private key
+     * @throws GeneralSecurityException
+     *             if we couldn't parse the string
+     */
+    private static PrivateKey getPrivateKeyFromString(final String key) throws GeneralSecurityException {
```

Comment:
any reason this is private? seems like I just have to re-implement this / copy it if I'm working with strings

Reply:
Probably better to create a constructor that takes a `String`.

Reply:
yes or that...